### PR TITLE
Track more precisely tuple to tuple assignments in sim

### LINKED ISSUE #666: Tuple desctructuring with `sim`
It seems that the `sim` tactic does not handle tuple assignments very accurately, here is a reduced example:

```
module T = {
  proc test(x: int): int = {
    var y: int;
    (x, y) <- (x, 0);
    return y;
  }
}.

equiv test_e: T.test ~ T.test : true ==> ={res}.
proof. proc. sim. qed.
```
I hoped that this proof would succeed, but `sim` produces the goal
```
forall &1 &2, true => ={x}
```

[Zulip discussion on the topic.](https://formosa-crypto.zulipchat.com/#narrow/channel/308381-EasyCrypt.3A-questions-.26-features/topic/sim.20with.20tuple.20destructuring/near/486793368)

### PR DESCRIPTION
**Pull request by cassiersg**
Fixes #666

### CODE REVIEW

**Comment by cassiersg** [comment]
@strub @bgregoir  here it is.

**Comment by cassiersg** [comment]
@bgregoir Unfortunately it seems that now that we introduce intermediate variables, `sim` fails again:

```
module T = {
  proc test(x: int): int = {
    var l: int;
    (l, x) <- (0, x);
    return l;
  }
}.

equiv t: T.test ~ T.test : true ==> res{1} = res{2}.
proof.
proc.
sim.
qed.
```

results in
```
forall &1 &2, true => ={ 0}
```
(where ` 0` is the new variable we introduced).

Do you see what's wrong with the code? Or do you have suggestions on how to debug this ?

**Comment by bgregoir** [comment]
I think the order of the list should be reversed (remember rev) : 
(mk_assigns lvs' es) @ (mk_assigns lvs (lvs2expr lvs')) -->
((mk_assigns lvs (lvs2expr lvs') @ (mk_assigns lvs' es)) 
or 
rev ((mk_assigns lvs' es) @ (mk_assigns lvs (lvs2expr lvs')))

**Comment by cassiersg** [comment]
Thanks that was the issue indeed. Now all gimli CT proof work.

**Comment by cassiersg** [comment]
Now you can critisize me (constructively) for using the variable name hack.

**Comment by bgregoir** [comment]
I would love to do that. Unfortunately I don't know how we can propose a nicer solution without an important modification of the code. We need to have the memories of the two current functions to be able to create fresh names properly.

**Comment by cassiersg** [comment]
I'm fine with leaving this as-is, if it is sound.

**Comment by bgregoir** [comment]
I think we should use this function : 
   let me, anames = EcMemory.bindall_fresh f.f_sig.fs_anames me in
like in ecPhlInline.ml.
But we need a memenv (in fact 2), we can discuss of this tomorrow.

**Review by bgregoir — APPROVED** [review]

File: src/phl/ecPhlEqobs.ml [hunk 1]
```diff
@@ -198,6 +198,23 @@ let rec s_eqobs_in_rev rsl rsr sim local (eqo:Mpv2.t) =
 
 and i_eqobs_in il ir sim local (eqo:Mpv2.t) =
   match il.i_node, ir.i_node with
+  | Sasgn(LvTuple lvls as lvl, ({e_node = Etuple els} as el)), Sasgn(LvTuple lvrs as lvr, ({e_node = Etuple ers} as er)) ->
+    let unpack lvs es =
+        let mk_assigns = List.map2 (fun lv e -> mk_instr (Sasgn (LvVar lv, e))) in
+        let lvs' = List.mapi (
+            (* These new variables will disappear after s_eqobs_in thanks
+               to the two consecutive assignments. *)
+            fun i (_, ty) -> PVloc (Format.sprintf " i_eqobs_in_tup_tmp%i" i), ty
```
Comment (strub) [inline src/phl/ecPhlEqobs.ml:207]: Using a non-syntaticaly valid variable name would ensure that you are collision free (e.g. by starting the name with a `$`)

Comment (cassiersg) [reply src/phl/ecPhlEqobs.ml:207]: We start with a space, that should be equivalent, right ? But maybe `$`  is more explicit ?

Comment (strub) [reply src/phl/ecPhlEqobs.ml:207]: Yes, please. Something more visual.

Comment (cassiersg) [reply src/phl/ecPhlEqobs.ml]: Changed ` ` to `$`.

**Pull request merged**